api/bugzilla/api_bugzilla.py

```diff
@@ -80,8 +145,141 @@ def __init__(self):
+        for n in names:
+            m = _CF_REL_RE.match(n)
+            if m:
+                releases.append((int(m.group(1)), n))
+        releases.sort(key=lambda t: t[0])  # sort by version number
+
+        if keep_last_n and len(releases) > keep_last_n:
+            releases = releases[-keep_last_n:]  # keep the 5 most recent trains
+
+        kept = [name for _, name in releases]
+        versions = [v for v, _ in releases]
+        print(f"[version-flags] last {len(kept)}: {kept} (versions={versions})")
+        return kept
+
+    def first_fixed_verified_by_version(self, history_rows):
+        """
+        Parse a bug's history and return {version:int -> first_timestamp: pd.Timestamp}
+        for the earliest time each cf_status_firefox{version} became fixed/verified.
+        """
+        out = {}
+        for when, fname, added, removed in history_rows:
+            m = _CF_REL_RE.match(fname or "")
+            if not m:
+                continue
+            if added in ("fixed", "verified"):
+                v = int(m.group(1))
+                # first occurrence only
+                if v not in out:
+                    out[v] = when
+        return out
+
+    def bugzilla_query_release_flags_for_tracked_bugs(
+            self, keep_last_n: int = 5, batch_size: int = 400, save_csv: bool = True):
+
+        version_fields = self._discover_release_status_fields(keep_last_n=keep_last_n)
+        if not version_fields:
+            return pd.DataFrame()
+
+        # Get status flags only for FIXED in DB
+        bug_ids = self.get_fixed_bug_ids()
+        if not bug_ids:
+            return pd.DataFrame()
+
+        include_fields = ["id", "cf_qa_whiteboard", "resolution", "keywords", "severity"] + version_fields
+        rows = []
+
+        for i in range(0, len(bug_ids), batch_size):
+            batch = bug_ids[i:i + batch_size]
+            query = {"id": ",".join(map(str, batch)), "include_fields": include_fields}
+            bugs = BugzillaHelper().query(query)
+
+            for bug in bugs:
+                for fname in version_fields:
+                    m = _CF_REL_RE.match(fname)
+                    if not m:
+                        continue
+                    version = int(m.group(1))
+                    raw = getattr(bug, fname, None)
+                    status = (str(raw).strip().lower() if raw and str(raw).strip() else '---') # noqa
+
+                    rows.append({
+                        "bugzilla_key": int(bug.id),
+                        "flag-version": version,
+                        "status": status,
+                        "keywords": ", ".join(bug.keywords),
+                        "severity": bug.severity,
+                        "qa-found-in": getattr(bug, "cf_qa_whiteboard", ""),
+                        "resolution": getattr(bug, "resolution", None)  # optional
+                    })
+
+        df = pd.DataFrame(rows)
+        if df.empty:
+            return df
```

Comment:
Inserting snapshots every run may grow the DB table quickly unless you deduplicate or prune; not sure where we want to handle that.